File: 735-asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();
        for(int asteroid : asteroids) {
            if(stack.isEmpty() || asteroid > 0 || (asteroid < 0 && stack.peek() < 0)) {
                stack.push(asteroid);
            }else{
                while(!stack.isEmpty() && stack.peek() > 0 && Math.abs(asteroid) > stack.peek()) {
                    stack.pop();
                }
                if(stack.isEmpty() || stack.peek() < 0) {
                    stack.push(asteroid);
                } else if (stack.peek() == Math.abs(asteroid)) {
                    stack.pop();
                }
            }
        }
        
        int[] result = new int[stack.size()];
        int index = stack.size() - 1;
        while(!stack.isEmpty()){
            result[index--] = stack.pop();
        }
        return result;
    }
}

File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids):
        result = []
        
        for asteroid in asteroids:
            if asteroid > 0:
                result.append(asteroid)
            else:
                # Destroy smaller positive asteroids
                while result and result[-1] > 0 and result[-1] < abs(asteroid):
                    result.pop()
                
                # Equal size collision
                if result and result[-1] == abs(asteroid):
                    result.pop()
                    continue
                
                # Add if no collision or safe
                if not result or result[-1] < 0:
                    result.append(asteroid)
        
        return result
